runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngine.kt

```diff
@@ -100,6 +103,40 @@ public class CrtHttpEngine(public val config: CrtHttpEngineConfig) : HttpClientE
         val stream = conn.makeRequest(engineRequest, respHandler)
         stream.activate()
 
+        if (request.isChunked) {
+            val bytesToRead = 65536L
+
+            when (request.body) {
+                is HttpBody.SourceContent -> {
+                    val source = (request.body as HttpBody.SourceContent).readFrom()
+                    val nextBuffer = SdkBuffer()
+                    var buffer = SdkBuffer()
+
+                    while (true) {
+                        source.read(buffer, bytesToRead)
+
+                        // read another set of bytes into a second buffer -- this is how we can "peek" to decide if this is the final chunk or not
+                        val isFinalChunk = source.read(nextBuffer, bytesToRead) == -1L
+
+                        stream.writeChunk(buffer.readToByteArray(), isFinalChunk)
+
+                        if (isFinalChunk) break else buffer = nextBuffer
+                    }
+                }
+                is HttpBody.ChannelContent -> {
+                    val chan = (request.body as HttpBody.ChannelContent).readFrom()
```

Comment:
I agree, but I tried it earlier and got this warning:

```
Smart cast to 'HttpBody.ChannelContent' is impossible, because 'request.body' is a public API property declared in different module
```

Reply:
The compiler _can't_ know this is safe because `body` is a `val` which, while always _read-only_, may be _non-constant_.